creusot/src/validate/terminates.rs

```diff
@@ -68,22 +68,42 @@ use rustc_trait_selection::traits::{
     normalize_param_env_or_error, specialization_graph, translate_args,
 };
 
+pub(crate) type RecursiveCalls = IndexMap<DefId, IndexSet<DefId>>;
+
 /// Validate that a `#[terminates]` function cannot loop indefinitely. This includes:
-/// - forbidding program function from using loops of any kind (this should be relaxed
-///   later).
+/// - forbidding program function from using loops without a variant.
 /// - forbidding (mutual) recursive calls, especially when traits are involved.
 ///
 /// Note that for logical functions, these are relaxed: we don't check loops, nor simple
 /// recursion, because why3 will handle it for us.
-pub(crate) fn validate_terminates(ctx: &TranslationCtx) {
+///
+/// # Returns
+///
+/// For each function with a `#[variant]`, returns the called function before
+/// which the variant should have decreased.
+/// FIXME: how do we ensure that the called function uses the _same_ variant?
```

Comment:
I don't see your point. Variants are terms depending on values of parameters. If there is a recursive call (possibly mutual), we check that the value of this term decreases. There is no notion of "same variant".

Reply:
I was thinking about having the same type here: I added a comment in the FIXME comment below.